cs-property-service/src/main/java/org/egov/cpt/service/RentDemandGenerationService.java

```diff
@@ -0,0 +1,147 @@
+		this.propertyRepository = propertyRepository;
+		this.producer = producer;
+		this.config = config;
+	}
+
+	public void createDemand(RentDemandCriteria demandCriteria) {
+		
+		PropertyCriteria propertyCriteria = new PropertyCriteria();
+		propertyCriteria.setRelations(new ArrayList<>());
+		List<Property> propertyList = propertyRepository.getProperties(propertyCriteria);
+
+		propertyList.forEach(property -> {
+			try {
+			propertyCriteria.setPropertyId(property.getId());
+			
+			List<RentDemand> rentDemandList = propertyRepository.getPropertyRentDemandDetails(propertyCriteria);
+			List<RentPayment> rentPaymentList = propertyRepository.getPropertyRentPaymentDetails(propertyCriteria);
+			RentAccount rentAccount = propertyRepository.getPropertyRentAccountDetails(propertyCriteria);
+			
+			if (!rentDemandList.isEmpty()) {
+
+				Comparator<RentDemand> compare = Comparator.comparing(RentDemand::getGenerationDate);
+				Optional<RentDemand> collectionDemand = rentDemandList.stream().min(compare);
+
+				int currentYear = LocalDate.now().getYear();
+				List<String> dateList = rentDemandList.stream().map(r -> r.getGenerationDate())
+						.map(date -> new Date(date).toInstant().atZone(ZoneId.systemDefault()).toLocalDate().toString())
+						.collect(Collectors.toList());
+
+				if (demandCriteria.getDate().isEmpty()) {
+					String currentDate = LocalDate.now().toString();
+
+					if (!dateList.contains(currentDate)) {
+						// generate demand
+						generateRentDemand(property, collectionDemand.get(), currentYear, rentDemandList, rentPaymentList, rentAccount);
+					}
+				} else {
+					String date = LocalDate.parse(demandCriteria.getDate(), FORMATTER).toString();
+					if (!dateList.contains(date)) {
+						// generate demand
+						generateRentDemand(property, collectionDemand.get(), currentYear, rentDemandList, rentPaymentList, rentAccount);
+						
+					}
+				}
+			} else {
+				log.debug("We are skipping generating rent demands for this property id: "+ property.getId() + " as there is no rent history");
+			}
+			} catch(Exception e) {
+				log.error("exception occured for property id: " + property.getId());
+			}
+		}
+				);
+
+	}
+
+	private void generateRentDemand(Property property, RentDemand collectionDemand, int currentYear,
+			List<RentDemand> rentDemandList, List<RentPayment> rentPaymentList, RentAccount rentAccount) {
+
+		int oldYear = new Date(collectionDemand.getGenerationDate()).toInstant().atZone(ZoneId.systemDefault())
+				.toLocalDate().getYear();
+
+		RentDemand rentDemand = new RentDemand();
+		String rendDemandId = UUID.randomUUID().toString();
+		rentDemand.setId(rendDemandId);
+		rentDemand.setPropertyId(property.getId());
+
+		rentDemand.setGenerationDate(LocalDateTime.now().atZone(ZoneId.systemDefault()).toInstant().toEpochMilli());
+
+		Double collectionPrincipal = collectionDemand.getCollectionPrincipal();
+		while (oldYear < currentYear) {
```

Comment:
This will not just depend on the previous year.
1. Some rented properties have 5 year increment cycle which mean that
When the first rent on Dec 2008 is 1000 and the increment percentage is 24 percent and increment period is 5 years, the rent will remain 1000 every month till Nov 2013, On Dec 2018 the rent becomes 1240.
2. Other properties have 1 year cycle which means that for a first time rent of Dec 2008 the next rent change will happen exactly on 1st Dec 2009 and not at the turn of the year like how it is being calculated here..